base/indices.jl

```diff
@@ -394,9 +394,52 @@ last(S::IdentityUnitRange) = last(S.indices)
 size(S::IdentityUnitRange) = (length(S.indices),)
 length(S::IdentityUnitRange) = length(S.indices)
 unsafe_length(S::IdentityUnitRange) = unsafe_length(S.indices)
-getindex(S::IdentityUnitRange, i::Int) = (@_inline_meta; @boundscheck checkbounds(S, i); i)
-getindex(S::IdentityUnitRange, i::AbstractUnitRange{<:Integer}) = (@_inline_meta; @boundscheck checkbounds(S, i); i)
-getindex(S::IdentityUnitRange, i::StepRange{<:Integer}) = (@_inline_meta; @boundscheck checkbounds(S, i); i)
+function getindex(S::IdentityUnitRange, i::Integer)
+    @_inline_meta
+    i isa Bool && throw(ArgumentError("invalid index: $i of type Bool"))
+    @boundscheck checkbounds(S, i)
+    i
```

Comment:
This might break indexing if the `IdentityUnitRange` wraps a `BigInt` `UnitRange` though. 

```julia
julia> IdentityUnitRange(1:big(2)^65)[end]
ERROR: InexactError: Int64(36893488147419103232)
Stacktrace:
 [1] Type
   @ ./gmp.jl:362 [inlined]
 [2] convert
   @ ./number.jl:7 [inlined]
 [3] to_index
   @ ./indices.jl:292 [inlined]
 [4] to_index
   @ ./indices.jl:277 [inlined]
 [5] to_indices
   @ ./indices.jl:333 [inlined]
 [6] to_indices
   @ ./indices.jl:330 [inlined]
 [7] getindex(S::IdentityUnitRange{UnitRange{BigInt}}, i::BigInt)
   @ Base ~/julia/base/indices.jl:400
 [8] top-level scope
   @ REPL[89]:1
```

Reply:
Sorry perhaps I've replied to the wrong comment, I meant using `to_indices`

Reply:
Yeah, but we declare the `IdentityUnitRange` as a `<: AbstractVector{Int}`.  Perhaps we should prevent construction of non-Int ranges or dynamically grab the eltype from the wrapped range.

Reply:
Does it make sense to define `IdentityUnitRange` as `IdentityUnitRange{T <:Integer, I <:AbstractUnitRange{T}} <: AbstractUnitRange{T}` instead? This way it'll work correctly for subtypes of `Integer`, and indexing may coerce to the type `T`. This is similar to how `IdOffsetRange` is implemented in `OffsetArrays`.

Reply:
Yes, that'd be best (and is precisely what I was meaning with that latter option).

Reply:
Although this appears to be breaking the precompile directives of `OffsetArrays` presently